src/main/java/org/opentripplanner/graph_builder/module/StreetLinkerModule.java

```diff
@@ -103,48 +103,89 @@ public void linkTransitStops(Graph graph, TransitModel transitModel) {
-        // street edges.
-        if (stopLocationsUsedForFlexTrips.contains(tStop.getStop())) {
-          modes = new TraverseModeSet(TraverseMode.WALK, TraverseMode.CAR);
-        }
+      // ordinarily stops only need to be accessible by foot
+      StopLinkType linkType = StopLinkType.WALK_ONLY;
+
+      if (
+        OTPFeature.FlexRouting.isOn() && stopLocationsUsedForFlexTrips.contains(tStop.getStop())
+      ) {
+        linkType = StopLinkType.WALK_AND_CAR;
       }
 
-      graph
-        .getLinker()
-        .linkVertexPermanently(
-          tStop,
-          modes,
-          LinkingDirection.BOTH_WAYS,
-          (vertex, streetVertex) ->
-            List.of(
-              StreetTransitStopLink.createStreetTransitStopLink(
-                (TransitStopVertex) vertex,
-                streetVertex
-              ),
-              StreetTransitStopLink.createStreetTransitStopLink(
-                streetVertex,
-                (TransitStopVertex) vertex
-              )
-            )
-        );
+      linkStopToStreetNetwork(tStop, linkType);
+
       //noinspection Convert2MethodRef
       progress.step(m -> LOG.info(m));
     }
     LOG.info(progress.completeMessage());
   }
 
+  /**
+   * Link a stop to the nearest "relevant" edges.
+   * <p>
+   * These are mostly walk edges but if a stop is used by a flex pattern it also needs to be
+   * car-accessible. Therefore, flex stops are ensured to be connected to the car-accessible
+   * edge. This may lead to several links being created.
+   */
+  private void linkStopToStreetNetwork(TransitStopVertex tStop, StopLinkType linkType) {
+    graph
+      .getLinker()
+      .linkVertexPermanently(
+        tStop,
+        WALK_ONLY,
+        LinkingDirection.BOTH_WAYS,
+        (walkVertex, walkStreetVertex) -> {
+          ArrayList<Edge> linkEdges = new ArrayList<>(
+            createStopLinkEdges((TransitStopVertex) walkVertex, walkStreetVertex)
+          );
+
+          // If regular stops or group stops are used for flex trips, they also need to be connected to car routable
+          // street edges. This does not apply to zones as street vertices store which zones they are part of.
+          if (
+            linkType == StopLinkType.WALK_AND_CAR && !walkStreetVertex.isConnectedToDriveableEdge()
+          ) {
+            graph
+              .getLinker()
+              .linkVertexPermanently(
+                tStop,
+                CAR_ONLY,
+                LinkingDirection.BOTH_WAYS,
+                (carVertex, carStreetVertex) -> {
+                  List<Edge> edges = createStopLinkEdges(
+                    (TransitStopVertex) carVertex,
+                    carStreetVertex
+                  );
+                  linkEdges.addAll(edges);
+                  return edges;
+                }
+              );
+          }
+
+          return linkEdges;
+        }
+      );
+  }
```

Comment:
The above function call has a large code-block as one of the arguments - from a clean code perspective this is not the "same level of abstraction". I would make a few private functions and inject those. You have some inline doc here as well witch then could be moved to be JavaDoc.

Reply:
I've reworked it a bit in the latest commit.